### ISSUE TITLE
DelayedCall should implement __lt__ to work with Python revisions [63827, 64116) See http://bugs.python.org/issue3051

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3359]
|<img alt="topia's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| topia reported|
|-|-|
|Trac ID|trac#3359|
|Type|defect|
|Created|2008-07-20 22:45:10Z|
|Branch|https://github.com/twisted/twisted/tree/delayedcall-ordering-3359|

heapq module changed to use less-than compare in Python trunk rev 63827.
but DelayedCall implements only ```__le__```, so reactor._pendingTimedCalls will be sorted with bad order.

After I wrote simple ```__lt__```, It works properly.

(See also: http://svn.python.org/view?rev=63827&view=rev)

Thanks.

Attachments:

* [twisted-ticket3359.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fdb/fdbc9d0b7a07312792a8377d7783d6d57549a156/f6272bcbc3ca615613fa82348830dcff0135c977.diff) (412 bytes) - added by topia on 2010-03-14 10:58:00Z - add __lt__ to DelayedCall
* [delayedcall-rich-comparison-3359-r28884.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fdb/fdbc9d0b7a07312792a8377d7783d6d57549a156/b6779669cf2a78573e499ecfa7c1cca1c98b913a.patch) (4627 bytes) - added by akira on 2010-04-30 18:03:49Z - added test cases
* [delayedcall-rich-comparison-3359-r28948.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fdb/fdbc9d0b7a07312792a8377d7783d6d57549a156/40c4fde58cd496f7a4ae4a0a4666266d23f162f4.patch) (6766 bytes) - added by akira on 2010-05-16 20:44:34Z - implemented suggestions from http://twistedmatrix.com/trac/ticket/3359#comment:10

<details><summary>Searchable metadata</summary>

```
trac-id__3359 3359
type__defect defect
reporter__topia topia
priority__lowest lowest
milestone__ 
branch__branches_delayedcall_ordering_3359 branches/delayedcall-ordering-3359
branch_author__akira akira
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1216593910000000 1216593910000000
changetime__1274108104000000 1274108104000000
version__None None
owner__ 
cc__therve cc__thijs cc__akira cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

The heapq module has been fixed to manage this case a few weeks ago. It should probably be done, but it's not critical (and current Twisted trunk works).

**Comment by twisted-trac** [comment]
|<img alt="topia's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|topia commented|
|-|-|

Actually, I'm using snapshot of Python trunk on 2008-06-07.
I will update my snapshot, the issue should be disappear, except using heapq's pure python implementations (as my knowledge, It's not critical, too.).

I'm sorry to report already resolved issues.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to topia|
|-|-|

Any chance you'd like to attach that simple `__lt__` as a patch, and submit it for review?

**Comment by twisted-trac** [comment]
|<img alt="topia's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|topia commented|
|-|-|

I attached the patch as 'twisted-ticket3359.diff'.
Could anyone review it?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

Does anyone know what version of Python is being referred to here? Would be good to update the ticket title and remove the reference to trunk.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

A narrow range of revisions between 2.6 and 2.7.  We complained upstream.  They fixed the behavior.  Strictly speaking, we don't **have** to fix this.  However, it's still probably a good idea to fully defined ordering for `DelayedCall`, rather than continuing to rely on just `__le__` being sufficient.

As far as the patch goes, it's a start.  I think we should define the rest of the methods as well, and add unit tests.

**Comment by twisted-trac** [comment]
|<img alt="akira's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|akira removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_10"></a>|@wallrj set owner to topia|
|-|-|

Akira,

The new tests make sense, but here are some comments (mostly coding standard issues):
 0. Add docstrings for all new tests
 0. Add a newsfile (http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles)
 0. Alter the ticket title as suggested in this comment: http://twistedmatrix.com/trac/ticket/3359#comment:7
 0. Add a more complete description of the "current inconsistent behaviour" and perhaps open a new ticket to fix it. OR
 0. make __le__ and __lt__ handle comparisons with other types - it looks like you can either fallback to default comparison or return NotImplemented if a comparison isn't possible
   * http://docs.python.org/reference/datamodel.html#basic-customization
   * http://www.voidspace.org.uk/python/articles/comparison.shtml
 0. Explain why assert(Not)Equals() isn't suitable
 0. In setUp I don't understand the id test - under what circumstances will that not be the case? Add an explanatory comment.
```
if id(self.zero) < id(self.one):
```
 0. In base.__le__ you say that "It is called for `>=` by CPython automatically" - what about other pythons? PyPy, Jython - maybe some further explanation.

**Comment by twisted-trac** [comment]
|<img alt="akira's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|akira removed owner|
|-|-|

Replying to [rwall](#note_10):
> The new tests make sense, but here are some comments (mostly coding standard issues):
>  0. Add docstrings for all new tests
Done. 

The utility of a docstring in an ordinary code is to provide all necessary information sufficient to use the code. Usually test method name is used as a docstring for a test (some test runners even replace test name by a docstring if it present for a test report). In most other use cases a programmer has source code of the test that should be self-explanatory and/or be commented appropriately. Therefore I don't understand usefulness of docstrings for tests if they are not used in test reporting (while running the tests).
>  0. Add a newsfile (http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles)
Done.
>  0. Alter the ticket title as suggested in this comment: http://twistedmatrix.com/trac/ticket/3359#comment:7
Done. 

Python implementations that use pure Python heapq module won't work from [r63827](http://svn.python.org/view?view=rev&revision=63827) onwards (CPython always uses C extension). [r64116](http://svn.python.org/view/python/trunk/Modules/_heapqmodule.c?r1=64048&r2=64116) affects only C implementation of heapq.
Python 3 also won't work. See http://bugs.python.org/issue3051 it uses LT instead of LE.
>  0. Add a more complete description of the "current inconsistent behaviour" and perhaps open a new ticket to fix it. OR
>  0. make __le__ and __lt__ handle comparisons with other types - it looks like you can either fallback to default comparison or return NotImplemented if a comparison isn't possible
By "inconsistent" I meant that comparison involving base.DelayedCall instances raises/doesn't raise AttributeError depending on argument order. 
I don't know what the correct behavior should be. Not raising AttributeError might mask a bug and is not *backwards-compatible*. On the other hand comparison operation should return NotImplemented if it can't perform the comparison for a given type.
>  0. Explain why assert(Not)Equals() isn't suitable
Done.
>  0. In setUp I don't understand the id test - under what circumstances will that not be the case? Add an explanatory comment.
> ```
> if id(self.zero) < id(self.one):
> ```
Done.
>  0. In base.__le__ you say that "It is called for `>=` by CPython automatically" - what about other pythons? PyPy, Jython - maybe some further explanation.

My remark was based on the quote (http://www.voidspace.org.uk/python/articles/comparison.shtml):

  Interestingly, IronPython has it a little better. You still have to provide both `__eq__` and `__ne__`, but you only need to provide `__lt__` or `__gt__`  as well; and then all comparisons give the right answer.

And [pep-0207](http://www.python.org/dev/peps/pep-0207/):

    If the object on the left side of the operator does not define an
    appropriate rich comparison operator (either at the C level or
    with one of the special methods, then the comparison is reversed,
    and the right hand operator is called with the opposite operator,
    and the two objects are swapped. This assumes that a < b and b > a
    are equivalent, as are a <= b and b >= a, and that == and != are
    commutative (e.g. a == b if and only if b == a).

I've removed the remark to avoid confusion.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [28951]) Branching to 'delayedcall-ordering-3359'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [28952]) Apply delayedcall-rich-comparison-3359-r28948.patch

refs [#3359](https://github.com/twisted/twisted/issues/3359)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [28953]) Move the DelayedCall tests into twisted/internet/test/

refs [#3359](https://github.com/twisted/twisted/issues/3359)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @exarkun|
|-|-|

> Therefore I don't understand usefulness of docstrings for tests if they are not used in test reporting (while running the tests).

They're for future maintainers to have to understand what the tests are supposed to be testing.

> On the other hand comparison operation should return NotImplemented if it can't perform the comparison for a given type. 

This ordering is basically an implementation detail, anyway.  Leaving the current behavior is fine.

I moved the tests to a newer test module where we're trying to consolidate tests for the `twisted.internet.base` module.  I also adjusted the docstrings to be a bit more specific, and removed some of the assertions for behavior we don't really care about.  The rest of this looks good, thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [28957]) Merge delayedcall-ordering-3359

Author: topia, akira, exarkun
Reviewer: rwall, exarkun
Fixes: [#3359](https://github.com/twisted/twisted/issues/3359)


Implement `<` for `DelayedCall` (which already implemented `<=`, which
by itself is sometimes not enough to guarantee sensible results).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
